### ISSUE TITLE
Hyperparameters in custom hyp.yaml Not Applied During Model Validation

### ISSUE THREAD

**Issue by adnan119** [ultralytics/ultralytics#18219]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Issue Description
I am currently experiencing an issue where custom hyperparameters specified in my custom_hyp.yaml file are not being applied during the validation process using the Ultralytics YOLO model. Despite providing a modified hyp_yaml_path, the validation still runs with default settings.Interestingly, these custom hyperparameters are correctly applied during the training phase when specified in the custom_hyp.yaml file.

### Expected Behavior
I expect the model to validate using the hyperparameters specified in the custom_hyp.yaml file.

### Actual Behavior
The model validation proceeds with the default hyperparameters, ignoring the settings specified in hyp.yaml. This issue is specific to the validation process as the training process correctly applies the same type of hyperparameter modifications.

I have tried explicitly setting the mode to "val" in the configuration to ensure that the validation mode recognizes the hyperparameters, but this has had no effect on the issue.

_**I am unsure if this is a bug or a usage error on my part. Any guidance on whether there's a specific method or additional steps required to ensure custom hyperparameters are applied during validation would be greatly appreciated.**_

### Environment

Ultralytics 8.3.35 🚀 Python-3.12.7 torch-2.5.1+cu124 CUDA:0 (NVIDIA A40, 45403MiB)
Setup complete ✅ (96 CPUs, 503.5 GB RAM, 4.1/20.0 GB disk)

OS                  Linux-6.5.0-45-generic-x86_64-with-glibc2.35
Environment         Docker
Python              3.12.7
Install             git
RAM                 503.51 GB
Disk                4.1/20.0 GB
CPU                 Intel Xeon Gold 6342 2.80GHz
CPU count           96
GPU                 NVIDIA A40, 45403MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 11.0.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.1>=1.8.0
torch               ✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1>=0.9.0
tqdm                ✅ 4.66.6>=4.64.0
psutil              ✅ 6.1.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.10>=2.0.0

### Minimal Reproducible Example

I'm running the validation using the following code snippet:

```

from ultralytics import YOLO
import torch

# Define paths and settings
train_yaml_path = 'path/to/train.yaml'
hyp_yaml_path = 'path/to/hyp.yaml'
RUNS_DIR = 'runs'
experiment_name = 'experiment1'
run_name = 'test_run'

# Load model and run validation
model = YOLO("yolov5s.pt")
results = model.val(
    data=train_yaml_path,
    cfg=hyp_yaml_path,
    split="test",
    project=f"{RUNS_DIR}/{experiment_name}",
    name=run_name,
    device= "cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "CPU",
)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @adnan119, thank you for your interest in Ultralytics 🚀!

We appreciate you taking the time to report this issue and for including such a detailed description of the problem and your setup. From your provided explanation, it seems like this may be a 🐛 bug. To help us investigate further, could you please ensure that you provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/) if one is not already included? An MRE will help confirm whether the issue lies in the code or your configuration.

In the meantime, here are a few steps to ensure you're running the latest version and setup:
1. **Upgrade**: Verify you're using the latest `ultralytics` release. Run the following command:
```bash
pip install -U ultralytics
```

2. **Dependencies**: Ensure all required dependencies are installed and up-to-date. You can check the complete list in the [requirements file](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml).

3. **Environment**: Consider running the validation in one of our preconfigured, verified environments for consistent results:
  - Notebooks with free GPU:
    <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
  - **Google Cloud**: See our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
  - **Amazon Web Services**: Use our [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
  - **Docker Image**: See our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls" /></a>.

4. **Validation Configuration**: Please double-check that the custom `hyp.yaml` path is correctly included in your `val()` method with proper field formatting.

If you’ve already upgraded and the issue persists, an Ultralytics engineer will assist you soon to help resolve this problem. While you wait, feel free to explore our [Documentation](https://docs.ultralytics.com/) for further insights into configurations and customizations.

For real-time discussions, join us on [Discord](https://discord.com/invite/ultralytics) 🎧, or visit the [Ultralytics Discourse](https://community.ultralytics.com) and [Subreddit](https://reddit.com/r/Ultralytics) for in-depth conversations with our community.

Thank you again for bringing this to our attention!

**Comment by Y-T-G** [comment]
What's inside the yaml file?

**Comment by adnan119** [comment]
It's the usual Hyperparameters and train/val/predict settings, I'm mostly trying to play around with how iou and conf. threshold would affect the outputs.

Here's how yaml file, the path of which gets passed into the hyp_yaml_path variable looks like for me:

```
# Ultralytics YOLO 🚀, AGPL-3.0 license
# Default training settings and hyperparameters for medium-augmentation COCO training

task: detect # (str) YOLO task, i.e. detect, segment, classify, pose, obb
mode: val # (str) YOLO mode, i.e. train, val, predict, export, track, benchmark

# Train settings -------------------------------------------------------------------------------------------------------
model: # (str, optional) path to model file, i.e. yolov8n.pt, yolov8n.yaml
data: # (str, optional) path to data file, i.e. coco8.yaml
epochs: 700 # (int) number of epochs to train for
time: # (float, optional) number of hours to train for, overrides epochs if supplied
patience: 100 # (int) epochs to wait for no observable improvement for early stopping of training
batch: 201 # (int) number of images per batch (-1 for AutoBatch)
imgsz: 640 # (int | list) input images size as int for train and val modes, or list[h,w] for predict and export modes
save: True # (bool) save train checkpoints and predict results
save_period: -1 # (int) Save checkpoint every x epochs (disabled if < 1)
cache: ram # (bool) True/ram, disk or False. Use cache for data loading
device: # (int | str | list, optional) device to run on, i.e. cuda device=0 or device=0,1,2,3 or device=cpu
workers: 8 # (int) number of worker threads for data loading (per RANK if DDP)
project: # (str, optional) project name
name: # (str, optional) experiment name, results saved to 'project/name' directory
exist_ok: False # (bool) whether to overwrite existing experiment
pretrained: True # (bool | str) whether to use a pretrained model (bool) or a model to load weights from (str)
optimizer: auto # (str) optimizer to use, choices=[SGD, Adam, Adamax, AdamW, NAdam, RAdam, RMSProp, auto]
verbose: True # (bool) whether to print verbose output
seed: 0 # (int) random seed for reproducibility
deterministic: True # (bool) whether to enable deterministic mode
single_cls: False # (bool) train multi-class data as single-class
rect: False # (bool) rectangular training if mode='train' or rectangular validation if mode='val'
cos_lr: False # (bool) use cosine learning rate scheduler
close_mosaic: 0 # (int) disable mosaic augmentation for final epochs (0 to disable)
resume: False # (bool) resume training from last checkpoint
amp: True # (bool) Automatic Mixed Precision (AMP) training, choices=[True, False], True runs AMP check
fraction: 1.0 # (float) dataset fraction to train on (default is 1.0, all images in train set)
profile: False # (bool) profile ONNX and TensorRT speeds during training for loggers
freeze: None # (int | list, optional) freeze first n layers, or freeze list of layer indices during training
multi_scale: False # (bool) Whether to use multiscale during training
# Segmentation
overlap_mask: True # (bool) merge object masks into a single image mask during training (segment train only)
mask_ratio: 4 # (int) mask downsample ratio (segment train only)
# Classification
dropout: 0.0 # (float) use dropout regularization (classify train only)

# Val/Test settings ----------------------------------------------------------------------------------------------------
val: True # (bool) validate/test during training
split: test # (str) dataset split to use for validation, i.e. 'val', 'test' or 'train'
save_json: False # (bool) save results to JSON file
save_hybrid: False # (bool) save hybrid version of labels (labels + additional predictions)
conf: 0.1 # (float, optional) object confidence threshold for detection (default 0.25 predict, 0.001 val)
iou: 0.5 # (float) intersection over union (IoU) threshold for NMS
max_det: 300 # (int) maximum number of detections per image
half: False # (bool) use half precision (FP16)
dnn: False # (bool) use OpenCV DNN for ONNX inference
plots: True # (bool) save plots and images during train/val

# Predict settings -----------------------------------------------------------------------------------------------------
source: # (str, optional) source directory for images or videos
vid_stride: 1 # (int) video frame-rate stride
stream_buffer: False # (bool) buffer all streaming frames (True) or return the most recent frame (False)
visualize: False # (bool) visualize model features
augment: False # (bool) apply image augmentation to prediction sources
agnostic_nms: False # (bool) class-agnostic NMS
classes: # (int | list[int], optional) filter results by class, i.e. classes=0, or classes=[0,2,3]
retina_masks: False # (bool) use high-resolution segmentation masks
embed: # (list[int], optional) return feature vectors/embeddings from given layers

# Visualize settings ---------------------------------------------------------------------------------------------------
show: False # (bool) show predicted images and videos if environment allows
save_frames: False # (bool) save predicted individual video frames
save_txt: False # (bool) save results as .txt file
save_conf: False # (bool) save results with confidence scores
save_crop: False # (bool) save cropped images with results
show_labels: True # (bool) show prediction labels, i.e. 'person'
show_conf: True # (bool) show prediction confidence, i.e. '0.99'
show_boxes: True # (bool) show prediction boxes
line_width: # (int, optional) line width of the bounding boxes. Scaled to image size if None.

# Export settings ------------------------------------------------------------------------------------------------------
format: torchscript # (str) format to export to, choices at https://docs.ultralytics.com/modes/export/#export-formats
keras: False # (bool) use Kera=s
optimize: False # (bool) TorchScript: optimize for mobile
int8: False # (bool) CoreML/TF INT8 quantization
dynamic: False # (bool) ONNX/TF/TensorRT: dynamic axes
simplify: True # (bool) ONNX: simplify model using `onnxslim`
opset: # (int, optional) ONNX: opset version
workspace: 4 # (int) TensorRT: workspace size (GB)
nms: False # (bool) CoreML: add NMS

# Hyperparameters ------------------------------------------------------------------------------------------------------
lr0: 0.006 # (float) initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
lrf: 0.01 # (float) final learning rate (lr0 * lrf)
momentum: 0.937 # (float) SGD momentum/Adam beta1
weight_decay: 0.0005 # (float) optimizer weight decay 5e-4
warmup_epochs: 0 # (float) warmup epochs (fractions ok)
warmup_momentum: 0.8 # (float) warmup initial momentum
warmup_bias_lr: 0.1 # (float) warmup initial bias lr
box: 7.5 # (float) box loss gain
cls: 0.5 # (float) cls loss gain (scale with pixels)
dfl: 1.5 # (float) dfl loss gain
pose: 12.0 # (float) pose loss gain
kobj: 1.0 # (float) keypoint obj loss gain
label_smoothing: 0.0 # (float) label smoothing (fraction)
nbs: 64 # (int) nominal batch size
hsv_h: 0.0 # (float) image HSV-Hue augmentation (fraction)
hsv_s: 0.0 # (float) image HSV-Saturation augmentation (fraction)
hsv_v: 0.2 # (float) image HSV-Value augmentation (fraction)
degrees: 45.0 # (float) image rotation (+/- deg)
translate: 0.1 # (float) image translation (+/- fraction)
scale: 0.0 # (float) image scale (+/- gain)
shear: 0.0 # (float) image shear (+/- deg)
perspective: 0.0 # (float) image perspective (+/- fraction), range 0-0.001
flipud: 0.0 # (float) image flip up-down (probability)
fliplr: 0.5 # (float) image flip left-right (probability)
bgr: 0.0 # (float) image channel BGR (probability)
mosaic: 0.0 # (float) image mosaic (probability)
mixup: 0.0 # (float) image mixup (probability)
copy_paste: 0.0 # (float) segment copy-paste (probability)
copy_paste_mode: "flip" # (str) the method to do copy_paste augmentation (flip, mixup)
auto_augment: randaugment # (str) auto augmentation policy for classification (randaugment, autoaugment, augmix)
erasing: 0.4 # (float) probability of random erasing during classification training (0-0.9), 0 means no erasing, must be less than 1.0.
crop_fraction: 1.0 # (float) image crop fraction for classification (0.1-1), 1.0 means no crop, must be greater than 0.

# Custom config.yaml ---------------------------------------------------------------------------------------------------
cfg: # (str, optional) for overriding defaults.yaml

# Tracker settings ------------------------------------------------------------------------------------------------------
tracker: botsort.yaml # (str) tracker type, choices=[botsort.yaml, bytetrack.yaml]
```

**Comment by glenn-jocher** [comment]
@adnan119 thank you for providing the detailed YAML configuration. Currently, the validation process in Ultralytics YOLO models does not automatically apply custom hyperparameters like `conf` and `iou` from the `hyp.yaml` file. These parameters need to be explicitly passed as arguments in the `model.val()` function.

For your case, you can modify your validation code like this:

```python
results = model.val(
    data=train_yaml_path,
    conf=0.1,  # Set your desired confidence threshold
    iou=0.5,   # Set your desired IoU threshold
    split="test",
    project=f"{RUNS_DIR}/{experiment_name}",
    name=run_name,
    device="cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "CPU",
)
```

This ensures the validation process uses the specified thresholds. Let me know if you have further questions!

**Comment by adnan119** [comment]
Thanks for the clarification, @glenn-jocher. However, I feel it would be better if we could also load these validation parameters from the hyp.yaml file. This would help keep everything in one place.

**Comment by glenn-jocher** [comment]
Thank you for the feedback! Currently, validation parameters need to be explicitly passed during the `val()` call, but I’ll share your suggestion with the team for consideration in future updates. For now, you can manage consistency by programmatically loading parameters from your `hyp.yaml` file into the `val()` function.
